ffsync/ordered_id.go

```diff
@@ -83,3 +91,143 @@ func (m *memoryIdGenerator) NextId(namespace string) (OrderedId, error) {
 	nextId := atomic.AddUint64((*uint64)(m.counters[namespace]), 1)
 	return (*Uint64OrderedId)(&nextId), nil
 }
+
+func NewETCDOrderedIdGenerator() (OrderedIdGenerator, error) {
+	etcdHost := helpers.GetEnv("ETCD_HOST", "localhost")
+	etcdPort := helpers.GetEnv("ETCD_PORT", "2379")
+
+	etcdURL := fmt.Sprintf("http://%s:%s", etcdHost, etcdPort)
+
+	client, err := clientv3.New(clientv3.Config{
+		Endpoints: []string{etcdURL},
+	})
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to create etcd client: %w", err))
+	}
+
+	session, err := concurrency.NewSession(client)
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to create etcd session: %w", err))
+	}
+
+	return &etcdIdGenerator{
+		client:  client,
+		session: session,
+		ctx:     context.Background(),
+	}, nil
+}
+
+type etcdIdGenerator struct {
+	client  *clientv3.Client
+	session *concurrency.Session
+	ctx     context.Context
+}
+
+func (etcd *etcdIdGenerator) NextId(namespace string) (OrderedId, error) {
+	if namespace == "" {
+		return nil, fferr.NewInternalError(fmt.Errorf("cannot generate ID for empty namespace"))
+	}
+
+	// Lock the namespace to prevent concurrent ID generation
+	lockKey := etcd_id_key + namespace
+	lockMutex := concurrency.NewMutex(etcd.session, lockKey)
+	if err := lockMutex.Lock(etcd.ctx); err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to lock key %s: %w", lockKey, err))
+	}
+	defer lockMutex.Unlock(etcd.ctx)
+
+	// Get the current ID
+	resp, err := etcd.client.Get(etcd.ctx, lockKey)
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to get key %s: %w", lockKey, err))
+	}
+
+	// Initialize the ID if it doesn't exist
+	var nextId uint64
+	if len(resp.Kvs) == 0 {
+		nextId = 1
+	} else {
+		if err := json.Unmarshal(resp.Kvs[0].Value, &nextId); err != nil {
+			return nil, fferr.NewInternalError(fmt.Errorf("failed to unmarshal ID: %w", err))
+		}
+		nextId++
+	}
+
+	// Update the ID
+	nextIdBytes, err := json.Marshal(nextId)
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to marshal ID: %w", err))
+	}
+	if _, err := etcd.client.Put(etcd.ctx, lockKey, string(nextIdBytes)); err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to set key %s: %w", lockKey, err))
+	}
+
+	return (*Uint64OrderedId)(&nextId), nil
+}
+
+func NewRDSOrderedIdGenerator() (OrderedIdGenerator, error) {
```

Comment:
This can be done without locking in RDS:
UPDATE counters SET value = value + 1 WHERE id = <your_id> RETURNING value;